Serialise empty model to XML

codecurve opened the pull request:
This is for #31

agarny commented:
A few remarks:
 - Shouldn't the ```.cpp```/```.h``` files generated by CodeSynthesis XSD be named ```cellml_1_2.cpp``` and ```cellml_1_2.h``` rather than ```Cellml_1_2.cpp``` and ```Cellml_1_2.h```, respectively (i.e. all lower case)? Maybe this can be done by renaming ```Cellml_1_2.xsd``` to ```cellml_1_2.xsd```?
 - In fact, why the reference to 1.2? Shouldn't we drop that reference altogether? Just wondering.
 - Comments: I can see some comments ending with ```.``` (e.g. https://github.com/codecurve/libcellml/commit/45aaaa9c4c54e880d8955a40afb676f9c2d41d97#diff-54c9d32c0cd4684d21e2a2a8ba949cb9R28) and others not (e.g. https://github.com/codecurve/libcellml/commit/45aaaa9c4c54e880d8955a40afb676f9c2d41d97#diff-54c9d32c0cd4684d21e2a2a8ba949cb9R29). Is that intentional? Might be good to be consistent?
 - Just wondering: is the code currently following the Google coding style (as I believe it has been agreed?)? Considering that some styles have been changed, I would think so. So, just want to make sure that it is the case.
 - Wrong indentation? https://github.com/codecurve/libcellml/commit/45aaaa9c4c54e880d8955a40afb676f9c2d41d97#diff-c2834ecd762e5cef4f316b2c5981f557R25.
 - Some ```using namespace``` are still being used. See https://github.com/codecurve/libcellml/commit/45aaaa9c4c54e880d8955a40afb676f9c2d41d97#diff-5421c51c47d2405f1f9d861d1ecb8b8fR26.
 - The ```auto``` keyword should only be used to avoid clutter (see http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#auto). I am not sure this is the case at https://github.com/codecurve/libcellml/commit/45aaaa9c4c54e880d8955a40afb676f9c2d41d97#diff-5421c51c47d2405f1f9d861d1ecb8b8fR33 for example?
 - Might want to create a CMake macro for things that are going to be common to both libCellML and its tests. A candidate for this would be https://github.com/codecurve/libcellml/commit/37f26017b1a5d822bc6528572f5d99232357a076.
 - BuildBot: it would seem that we should upgrade the version of CMake to 3.2 from 3.1.1 on the Ubuntu VM.
 - BuildBot: need to resolve issues of unresolved external symbols.

codecurve commented:
Thanks @agarny 
I've renamed to cellml_1_2.*, as you suggested (i.e. starting on lowercase) (see a3574431a0a2446b9004f7afa84dc2a0f3023de0).
I think we plan to support multiple versions of CellML, so it will be necessary to distinguish these, hence the _1_2 for v1.2 is necessary (IMHO).
Yes, I am attempting to follow the Google style guide.
I've changed the use of `auto` that you highlighted to explicitly stating the type, and also to using a const reference, which is more correct, thanks for suggesting that.  (`const auto&` would have probably been more correct if using `auto`.)  (see 2154983657bb03e9a7659aa60ad131d2e9483f74)
Last week I asked @hsorby to upgrade the Ubuntu cmake to v3.2, I think he might do it today.
I've eliminated use of `using namespace` directives (see 4b12cc5e78119333da32ecfedb877efb9f50e580)
I've made the comments consistently end with full stops (see c4431f006f3692dd3be570edba84ad4248675727).
I'm waiting for @hsorby for access to the Windows Build slave to try and work out why it has unresolved symbols during linking, since it links fine on my Windows box.
The "wrong indentation" is required in order to use the raw string literals.  For the tests I did when I previously made quite a bit of progress serialising CellML v1.2 models with components, units and variables, I used raw string literals for the expected XML strings, and it is far more readable (see https://github.com/codecurve/libcellml/blob/e7a0fc5fe60be981f82da481710c05d8a473babb/tests/XmlBindingsSerialisationTest.cpp#L14-L21 for example)

codecurve commented:
@agarny Good catch, thanks, I've addressed that now. (See cf391bb3ee861cd2ac10bf285051f5608ac4964f).
I guess it is a matter of taste.  My personal guideline is to try and reduce clutter, so if I'm only resolving scope once, no `using` is needed, but more than once seems to reduce clutter, IMHO.

codecurve commented:
I've managed to reproduce the failed Windows link on my Windows PC by setting the `LIBCELLML_ENABLE_SHARED` cmake flag.  I'm trying @agarny's suggestion as shown here:
https://github.com/opencor/opencor/blob/master/src/plugins/miscellaneous/Compiler/src/compilerglobal.h#L25
and here
https://github.com/opencor/opencor/blob/master/src/plugins/miscellaneous/Compiler/src/compilerengine.h#L55

codecurve commented:
Windows build now working, see 89b2f09aa08ff33db45ce131a9634369156217f1.

agarny commented:
I seem to remember reading somewhere that there is a need for CMake 3.2, so how come everything is building fine using CMake 3.1.1 on Ubuntu (through BuildBot)?

Otherwise, I can see that there is a warning on the Ubuntu build: http://autotest.bioeng.auckland.ac.nz/libcellml-buildbot/builders/Unit%20Tests%20Builder%20Ubuntu%2014.04%2064%20Bit/builds/86/steps/compile/logs/stdio. I don't know what has been agreed, if anything, regarding warnings, but personally I would consider them as errors (i.e. build libCellML using ```/WX``` or ```-Werror``` on Windows and Linux/OS X, respectively).

codecurve commented:
Ubuntu cmake 3.1.1 seems to do mostly the correct thing, but 3.2 is required on Windows and OS X.
I've eliminated the warning, see 799e632.
As for setting the flags so warnings are treated as errors, I'm neutral, but I do think that is a separate issue and pull request, so I'll add an issue for it.

codecurve commented:
I've added #50 regarding the "compile warnings as errors" question.

codecurve commented:
"CMake macro for things that are going to be common" - done, see 54de041.  
#50 - done, as discussed with @agarny we decided to tack it onto this one, rather than make a separate pull request.

jonc125 commented:
On the `using` question, if a particular term is being used multiple times within a single function then it's reasonable to start the function with a `using ns1::ns2::term` for instance. The downside of doing so at a whole file level is that it's less clear what's in scope, and hence harder to read & reuse code. A `using namespace` declaration is even worse from this perspective, and even at a function level can lead to issues from scope pollution.

hsorby commented:
I'm not a big fan of this approach with regards to the bulk setting of the compiler flags.  I think this should be set on a per target basis.

The conditional tests for compiler options should be based on the compiler and not the platform.

This is no longer needed:

# To work around a bug conditionally set the CXX_STANDARD property
if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
  set_target_properties(cellml PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED ON)
endif()

I don't understand why this header guard has been changed to this:

#ifndef SRC_API_LIBCELLML_VERSION_H_

I would also set the CMAKE_MODULE_PATH and then just include the common.cmake file.  

The name of the macro is misleading it really should be set_compiler_flags, and as I said earlier this should be done on per target basis.

codecurve commented:
The header guard is in the format required by the Google style guide as I understand it: http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#The__define_Guard

hsorby commented:
The document string in this line isn't correct.

set( CMAKE_CODEBLOCKS_EXECUTABLE "" CACHE INTERNAL "Internalise BUILD_SHARED_LIBS" FORCE )

hsorby commented:
Then I don't think you understand it.  It should be according to the style guide

LIBCELLML_API_LIBCELLML_VERSION_H_  

or 

LIBCELLML_LIBCELLML_VERSION_H_

depending on how you define the source tree.  As this is an API header file that gets installed I would say the second version is the correct one according to the standard.

codecurve commented:
Cpplint looks for .git, and infers `SRC_API_LIBCELLML_VERSION_H_`, I haven’t figured out how to override it yet from within the editor.
I really don’t mind what convention we use, so long as it functions correctly (i.e. prevents duplicate inclusion), happy with either of your suggestions, currently opting for the shorter of your suggestions.

codecurve commented:
See bcbb61b0f902b05adce42bd037a3c03f3ad72ff7

codecurve commented:
You committed this, do you know what the correct string should be? (see ccd7a4f)

codecurve commented:
Done, see 7dcbb9b82be1095d4416cfca1b0fdbd138f0ca73 (re part of comment, quoted below).

The name of the macro is misleading it really should be set_compiler_flags, and as I said earlier this should be done on per target basis.

codecurve commented:
I need some help, please, with using the CMAKE_MODULE_PATH proposal, for some reason, the numerous ways I’ve tried don’t work.

codecurve commented:
All of the items in this comment have now been addressed.

Original comment:


I'm not a big fan of this approach with regards to the bulk setting of the compiler flags. I think this should be set on a per target basis.

The conditional tests for compiler options should be based on the compiler and not the platform.

This is no longer needed:

To work around a bug conditionally set the CXX_STANDARD property

if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
set_target_properties(cellml PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED ON)
endif()

I don't understand why this header guard has been changed to this:

#ifndef SRC_API_LIBCELLML_VERSION_H_

I would also set the CMAKE_MODULE_PATH and then just include the common.cmake file.

The name of the macro is misleading it really should be set_compiler_flags, and as I said earlier this should be done on per target basis.

nickerso commented:
@codecurve it would be good to make use of quoting when you are quoting previous comments. Also, this bit:

>  I said earlier this should be done on per target basis.

doesn't seem to have been addressed?

codecurve commented:
As I understand it, the "per target basis” issue has been addressed, the cmake file for the library target uses the macro independently from the cmake file for the tests target using the macro.

nickerso commented:
I suspect what @hsorby meant was that rather than setting the global C flags you should be setting properties on each target. Then there is no need to reset the global flags in different parts of the source tree.

codecurve commented:
Had a chat with @hsorby, he will send through more info.

codecurve commented:
I've reviewed commits c3f756c, 3be776a and 891d2a1, they look good to me.
4017ed0 was discussed with @hsorby this morning, and has been done as I understand he intended, so I think this pull request should be merged now.

agarny commented:
For 4017ed0, you might want to rewrite as follows:

```
   if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
     set(_COMPILER_WAE -Wall -W -Werror)
   elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
     set(_COMPILER_WAE -Wall -W -Werror)
   elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
     set(_COMPILER_WAE /W3)
   endif()
```

This aside, it looks fine to me.

codecurve commented:
Done, see 7483eb711b78c358f6d163901cafe25240152bd8 .  Thanks for the suggestion.

From: Alan Garny <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cellml/libcellml <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Wednesday, 1 April 2015 3:17 pm
To: cellml/libcellml <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com>>
Cc: Randall Britten <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz>>
Subject: Re: [libcellml] Serialise empty model to XML (#49)


if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
     set(_COMPILER_WAE -Wall -W -Werror)
   elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
     set(_COMPILER_WAE -Wall -W -Werror)
   elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
     set(_COMPILER_WAE /W3)
   endif()

hsorby commented:
As a further alternative it might be even better to:

   set(_GNU_FLAGS_COMPILER_COMPAT "GNU" "AppleClang" "Clang")
   list(FIND _GNU_FLAGS_COMPILER_COMPAT "${CMAKE_CXX_COMPILER_ID}" _INDEX)
   if(${_INDEX} GREATER -1)
     set(_COMPILER_WAE -Wall -W -Werror)
   elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
     set(_COMPILER_WAE /W3)
   endif()


On 04/01/15 15:28, codecurve wrote:
> Done, see 7483eb711b78c358f6d163901cafe25240152bd8 . Thanks for the 
> suggestion.
>
> From: Alan Garny 
> <notifications@github.com<mailto:notifications@github.com>>
> Reply-To: cellml/libcellml 
> <reply@reply.github.com<mailto:reply@reply.github.com>>
> Date: Wednesday, 1 April 2015 3:17 pm
> To: cellml/libcellml 
> <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com>>
> Cc: Randall Britten 
> <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz>>
> Subject: Re: [libcellml] Serialise empty model to XML (#49)
>
>
> if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
> set(_COMPILER_WAE -Wall -W -Werror)
> elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
> set(_COMPILER_WAE -Wall -W -Werror)
> elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
> set(_COMPILER_WAE /W3)
> endif()
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cellml/libcellml/pull/49#issuecomment-88317649>.
>

codecurve commented:
Done, see b603ca1

From: Hugh Sorby <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cellml/libcellml <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Wednesday, 1 April 2015 3:32 pm
To: cellml/libcellml <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com>>
Cc: Randall Britten <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz>>
Subject: Re: [libcellml] Serialise empty model to XML (#49)

As a further alternative it might be even better to:

set(_GNU_FLAGS_COMPILER_COMPAT "GNU" "AppleClang" "Clang")
list(FIND _GNU_FLAGS_COMPILER_COMPAT "${CMAKE_CXX_COMPILER_ID}" _INDEX)
if(${_INDEX} GREATER -1)
set(_COMPILER_WAE -Wall -W -Werror)
elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
set(_COMPILER_WAE /W3)
endif()


On 04/01/15 15:28, codecurve wrote:
> Done, see 7483eb711b78c358f6d163901cafe25240152bd8 . Thanks for the
> suggestion.
>
> From: Alan Garny
> <notifications@github.com<mailto:notifications@github.com><mailto:notifications@github.com>>
> Reply-To: cellml/libcellml
> <reply@reply.github.com<mailto:reply@reply.github.com><mailto:reply@reply.github.com>>
> Date: Wednesday, 1 April 2015 3:17 pm
> To: cellml/libcellml
> <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com><mailto:libcellml@noreply.github.com>>
> Cc: Randall Britten
> <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz><mailto:r.britten@auckland.ac.nz>>
> Subject: Re: [libcellml] Serialise empty model to XML (#49)
>
>
> if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
> set(_COMPILER_WAE -Wall -W -Werror)
> elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
> set(_COMPILER_WAE -Wall -W -Werror)
> elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
> set(_COMPILER_WAE /W3)
> endif()
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cellml/libcellml/pull/49#issuecomment-88317649>.
>


—
Reply to this email directly or view it on GitHub<https://github.com/cellml/libcellml/pull/49#issuecomment-88317974>.

agarny commented:
Fine by me.

codecurve commented:
20 hours have passed since all review items have been addressed (as far as I can tell). I think this can be merged, please.

hsorby commented:
the header file ${CMAKE_CURRENT_SOURCE_DIR}/cellml_1_2.h is listed as an API header file but this is not the case

hsorby commented:
the ${CMAKE_INSTALL_PREFIX}/ is not required in the install(FILES command

codecurve commented:
Done, thanks.  See 6896bff.

From: Hugh Sorby <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cellml/libcellml <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Thursday, 2 April 2015 1:11 pm
To: cellml/libcellml <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com>>
Cc: Randall Britten <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz>>
Subject: Re: [libcellml] Serialise empty model to XML (#49)


the header file ${CMAKE_CURRENT_SOURCE_DIR}/cellml_1_2.h is listed as an API header file but this is not the case

—
Reply to this email directly or view it on GitHub<https://github.com/cellml/libcellml/pull/49#issuecomment-88672213>.

codecurve commented:
Done, see f7bcb5be8c09fb9748e0cfe6a360a81e5d7619c1

From: Hugh Sorby <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cellml/libcellml <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Thursday, 2 April 2015 1:13 pm
To: cellml/libcellml <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com>>
Cc: Randall Britten <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz>>
Subject: Re: [libcellml] Serialise empty model to XML (#49)


the ${CMAKE_INSTALL_PREFIX}/ is not required in the install(FILES command

—
Reply to this email directly or view it on GitHub<https://github.com/cellml/libcellml/pull/49#issuecomment-88672330>.

hsorby commented:
I have created another pull request onto @codecurve empty-model-01 branch with the changes to the CMake related files that make me happy.

A few other  points:
 - I would think the generated files should be in a directory named 'generated' under the source tree
 - the xsd, generate.sh script and xsd-options.txt files placed in a 'utils' directory as a sibling of the src directory
 - the type-naming and function-naming options for the generation be set to 'knr' and not 'java'

codecurve commented:
Done, see 49051a61cd72e57fcb251865d111102bd574c0d5 and 367704734774b9b33524f5a91c645ad6199153ee


From: Hugh Sorby <notifications@github.com<mailto:notifications@github.com>>
Reply-To: cellml/libcellml <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Thursday, 2 April 2015 2:27 pm
To: cellml/libcellml <libcellml@noreply.github.com<mailto:libcellml@noreply.github.com>>
Cc: Randall Britten <r.britten@auckland.ac.nz<mailto:r.britten@auckland.ac.nz>>
Subject: Re: [libcellml] Serialise empty model to XML (#49)


I have created another pull request onto @codecurve<https://github.com/codecurve> empty-model-01 branch with the changes to the CMake related files that make me happy.

A few other points:

  *   I would think the generated files should be in a directory named 'generated' under the source tree
  *   the xsd, generate.sh script and xsd-options.txt files placed in a 'utils' directory as a sibling of the src directory
  *   the type-naming and function-naming options for the generation be set to 'knr' and not 'java'

—
Reply to this email directly or view it on GitHub<https://github.com/cellml/libcellml/pull/49#issuecomment-88682019>.

codecurve commented:
I chatted with @hsorby, and we agreed that keeping XSD `type-naming` and `function-naming` set to `java` is the best option for generated code, `knr` is too different from our target style, and it is not worth the cost to customise generation to exactly match our target style.

  *   the type-naming and function-naming options for the generation be set to 'knr' and not 'java'

codecurve commented:
Buildbot fail was not due to any problem in code, but due to download problem.
Passed on Window and Linux slaves, and builds and tests pass on my mac.
Sadly, there is no way yet to trigger a retry of a given build.

nickerso commented:
HI Randall, we can manually trigger a changeset to be redelivered - just ask. In this case though, the download problem for the mac machine is still there.

codecurve commented:
I have tried to access the Mac build slave that is having trouble with the download, but either my password has been changed, or I forgot it.  Anyway, I have a mirror branch in my repo (https://github.com/codecurve/libcellml/tree/empty-model-0-travis-on-02), which has Travis enabled, and as you can see, both Mac and Linux builds passing there.  See https://travis-ci.org/codecurve/libcellml/builds/57134328
